File: elixir-shyryan/lib/token.ex
defmodule Monkey.Token do
  @moduledoc """
  Token struct for the Monkey lang.
  """

  @type token_type ::
    :illegal
    | :eof
    | :ident
    | :int
    | :assign
    | :plus
    | :minus
    | :asterisk
    | :slash
    | :bang
    | :equal_equal
    | :not_equal
    | :greater_than
    | :less_than
    | :comma
    | :semicolon
    | :lparen
    | :rparen
    | :lsquirly
    | :rsquirly
    | :fn
    | :let
    | true
    | false
    | :if
    | :else
    | :return

  @type t :: %__MODULE__{
    type: token_type(),
    literal: String.t()
  }

  @enforce_keys [:type, :literal]
  defstruct [:type, :literal]

  @keywords %{
    "fn" => :fn,
    "let" => :let,
    "true" => true,
    "false" => false,
    "if" => :if,
    "else" => :else,
    "return" => :return
  }

  @literals %{
    illegal: "ILLEGAL",
    eof: "EOF",
    # identifiers and literals
    ident: "IDENT",
    # 123
    int: "INT",
    # operators
    assign: "=",
    plus: "+",
    minus: "-",
    asterisk: "*",
    slash: "/",
    bang: "!",
    equal_equal: "==",
    not_equal: "!=",
    greater_than: ">",
    less_than: "<",
    # delimiters
    comma: ",",
    semicolon: ",",
    lparen: "(",
    rparen: ")",
    lsquirly: "{",
    rsquirly: "}",
    fn: "FUNCTION",
    let: "LET",
    true: "TRUE",
    false: "FALSE",
    if: "IF",
    else: "ELSE",
    return: "RETURN"
  }

  @token_types Map.keys(@literals)

  @doc """
  Create a new token struct.
  """
  def new(type) when type in @token_types do
    %__MODULE__{type: type, literal: @literals[type]}
  end

  @doc """
  Lookup an identifier.
  """
  @spec lookup_identifier(String.t()) :: atom()
  def lookup_identifier(identifier) do
    Map.get(@keywords, identifier, :ident)
  end
end

File: csharp2/Monkey/Token.cs
namespace Monkey;

enum TokenType
{
    Illegal,
    Eof,

    // Identifiers
    Ident,
    Int,

    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    LessThan,
    GreaterThan,
    Equal,
    NotEqual,

    // Delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LSquirly,
    RSquirly,

    // Keywords
    Function,
    Let,
    If,
    Else,
    Return,
    True,
    False,
}

readonly record struct Token(TokenType Type, string? Literal = null)
{
    internal static Token Illegal = new(TokenType.Illegal, "ILLEGAL");
    internal static Token Eof = new(TokenType.Eof, "EOF");

    internal static Token Assign = new(TokenType.Assign, "=");
    internal static Token Plus = new(TokenType.Plus, "+");
    internal static Token Minus = new(TokenType.Minus, "-");
    internal static Token Bang = new(TokenType.Bang, "!");
    internal static Token Asterisk = new(TokenType.Asterisk, "*");
    internal static Token Slash = new(TokenType.Slash, "/");
    internal static Token LessThan = new(TokenType.LessThan, "<");
    internal static Token GreaterThan = new(TokenType.GreaterThan, ">");
    internal static Token Equal = new(TokenType.Equal, "==");
    internal static Token NotEqual = new(TokenType.NotEqual, "!=");

    internal static Token Comma = new(TokenType.Comma, ",");
    internal static Token Semicolon = new(TokenType.Semicolon, ";");
    internal static Token LParen = new(TokenType.LParen, "(");
    internal static Token RParen = new(TokenType.RParen, ")");
    internal static Token LSquirly = new(TokenType.LSquirly, "{");
    internal static Token RSquirly = new(TokenType.RSquirly, "}");

    internal static Token Function = new(TokenType.Function, "fn");
    internal static Token Let = new(TokenType.Let, "let");
    internal static Token If = new(TokenType.If, "if");
    internal static Token Else = new(TokenType.Else, "else");
    internal static Token Return = new(TokenType.Return, "return");
    internal static Token True = new(TokenType.True, "true");
    internal static Token False = new(TokenType.False, "false");
}
